File: CloudApi/src/main/java/com/proxy/service/api/log/tree/TreeGroup.java
package com.proxy.service.api.log.tree;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.proxy.service.api.log.base.LogTree;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * @author: cangHX
 * @data: 2024/11/15 09:59
 * @desc:
 */
public class TreeGroup extends LogTree {

    private static final LogTree[] TREE_ARRAY_EMPTY = new LogTree[0];
    private static final List<LogTree> FOREST = new ArrayList<>();
    private volatile static LogTree[] forestAsArray = TREE_ARRAY_EMPTY;

    /**
     * @noinspection ToArrayCallWithZeroLengthArrayArgument
     */
    public static void plant(@NonNull LogTree... trees) {
        synchronized (FOREST) {
            Collections.addAll(FOREST, trees);
            forestAsArray = FOREST.toArray(new LogTree[FOREST.size()]);
        }
    }

    @Override
    public void setTag(@NonNull String tag) {
        LogTree[] forest = forestAsArray;
        for (LogTree logTree : forest) {
            logTree.setTag(tag);
        }
    }

    @Override
    public void v(@NonNull String message, @NonNull Object[] args) {
        LogTree[] forest = forestAsArray;
        for (LogTree logTree : forest) {
            logTree.v(message, args);
        }
    }

    @Override
    public void v(@NonNull Throwable throwable, @NonNull String message, @NonNull Object... args) {
        LogTree[] forest = forestAsArray;
        for (LogTree logTree : forest) {
            logTree.v(throwable, message, args);
        }
    }

    @Override
    public void v(@NonNull Throwable throwable) {
        LogTree[] forest = forestAsArray;
        for (LogTree logTree : forest) {
            logTree.v(throwable);
        }
    }

    @Override
    public void d(@NonNull String message, @NonNull Object... args) {
        LogTree[] forest = forestAsArray;
        for (LogTree logTree : forest) {
            logTree.d(message, args);
        }
    }

    @Override
    public void d(@NonNull Throwable throwable, @NonNull String message, @NonNull Object... args) {
        LogTree[] forest = forestAsArray;
        for (LogTree logTree : forest) {
            logTree.d(throwable, message, args);
        }
    }

    @Override
    public void d(@NonNull Throwable throwable) {
        LogTree[] forest = forestAsArray;
        for (LogTree logTree : forest) {
            logTree.d(throwable);
        }
    }

    @Override
    public void i(@NonNull String message, @NonNull Object[] args) {
        LogTree[] forest = forestAsArray;
        for (LogTree logTree : forest) {
            logTree.i(message, args);
        }
    }

    @Override
    public void i(@NonNull Throwable throwable, @NonNull String message, @NonNull Object... args) {
        LogTree[] forest = forestAsArray;
        for (LogTree logTree : forest) {
            logTree.i(throwable, message, args);
        }
    }

    @Override
    public void i(@NonNull Throwable throwable) {
        LogTree[] forest = forestAsArray;
        for (LogTree logTree : forest) {
            logTree.i(throwable);
        }
    }

    @Override
    public void w(@NonNull String message, @NonNull Object[] args) {
        LogTree[] forest = forestAsArray;
        for (LogTree logTree : forest) {
            logTree.w(message, args);
        }
    }

    @Override
    public void w(@NonNull Throwable throwable, @NonNull String message, @NonNull Object... args) {
        LogTree[] forest = forestAsArray;
        for (LogTree logTree : forest) {
            logTree.w(throwable, message, args);
        }
    }

    @Override
    public void w(@NonNull Throwable throwable) {
        LogTree[] forest = forestAsArray;
        for (LogTree logTree : forest) {
            logTree.w(throwable);
        }
    }

    @Override
    public void e(@NonNull String message, @NonNull Object[] args) {
        LogTree[] forest = forestAsArray;
        for (LogTree logTree : forest) {
            logTree.e(message, args);
        }
    }

    @Override
    public void e(@NonNull Throwable throwable, @NonNull String message, @NonNull Object... args) {
        LogTree[] forest = forestAsArray;
        for (LogTree logTree : forest) {
            logTree.e(throwable, message, args);
        }
    }

    @Override
    public void e(@NonNull Throwable throwable) {
        LogTree[] forest = forestAsArray;
        for (LogTree logTree : forest) {
            logTree.e(throwable);
        }
    }

    @Override
    public void wtf(@NonNull String message, @NonNull Object[] args) {
        LogTree[] forest = forestAsArray;
        for (LogTree logTree : forest) {
            logTree.wtf(message, args);
        }
    }

    @Override
    public void wtf(@NonNull Throwable throwable, @NonNull String message, @NonNull Object... args) {
        LogTree[] forest = forestAsArray;
        for (LogTree logTree : forest) {
            logTree.wtf(throwable, message, args);
        }
    }

    @Override
    public void wtf(@NonNull Throwable throwable) {
        LogTree[] forest = forestAsArray;
        for (LogTree logTree : forest) {
            logTree.wtf(throwable);
        }
    }

    @Override
    public void log(int priority, @NonNull String message, @NonNull Object... args) {
        LogTree[] forest = forestAsArray;
        for (LogTree logTree : forest) {
            logTree.log(priority, message, args);
        }
    }

    @Override
    public void log(int priority, @NonNull Throwable throwable, @NonNull String message, @NonNull Object... args) {
        LogTree[] forest = forestAsArray;
        for (LogTree logTree : forest) {
            logTree.log(priority, throwable, message, args);
        }
    }

    @Override
    public void log(int priority, @NonNull Throwable throwable) {
        LogTree[] forest = forestAsArray;
        for (LogTree logTree : forest) {
            logTree.log(priority, throwable);
        }
    }

    @Override
    protected void onLog(int priority, @NonNull String tag, @NonNull String message, @Nullable Throwable throwable) {
        //nothing
    }
}

File: CloudApi/src/main/java/com/proxy/service/api/log/tree/TreeGroup.kt
package com.proxy.service.api.log.tree

import com.proxy.service.api.log.base.LogTree

/**
 * @author: cangHX
 * @data: 2024/11/15 09:59
 * @desc:
 */
class TreeGroup : LogTree() {

    companion object {
        private val TREE_ARRAY_EMPTY: Array<LogTree> = emptyArray()
        private val FOREST: ArrayList<LogTree> = ArrayList()

        @Volatile
        private var forestAsArray: Array<LogTree> = TREE_ARRAY_EMPTY

        fun plant(tree: LogTree) {
            synchronized(FOREST) {
                FOREST.add(tree)
                forestAsArray = FOREST.toTypedArray()
            }
        }
    }

    override fun setTag(tag: String) {
        forestAsArray.forEach {
            it.setTag(tag)
        }
    }

    override fun v(message: String, vararg args: Any) {
        forestAsArray.forEach {
            it.v(message, args)
        }
    }

    override fun v(throwable: Throwable, message: String, vararg args: Any) {
        forestAsArray.forEach {
            it.v(throwable, message, args)
        }
    }

    override fun v(throwable: Throwable) {
        forestAsArray.forEach {
            it.v(throwable)
        }
    }

    override fun d(message: String, vararg args: Any) {
        forestAsArray.forEach {
            it.d(message, args)
        }
    }

    override fun d(throwable: Throwable, message: String, vararg args: Any) {
        forestAsArray.forEach {
            it.d(throwable, message, args)
        }
    }

    override fun d(throwable: Throwable) {
        forestAsArray.forEach {
            it.d(throwable)
        }
    }

    override fun i(message: String, vararg args: Any) {
        forestAsArray.forEach {
            it.i(message, args)
        }
    }

    override fun i(throwable: Throwable, message: String, vararg args: Any) {
        forestAsArray.forEach {
            it.i(throwable, message, args)
        }
    }

    override fun i(throwable: Throwable) {
        forestAsArray.forEach {
            it.i(throwable)
        }
    }

    override fun w(message: String, vararg args: Any) {
        forestAsArray.forEach {
            it.w(message, args)
        }
    }

    override fun w(throwable: Throwable, message: String, vararg args: Any) {
        forestAsArray.forEach {
            it.w(throwable, message, args)
        }
    }

    override fun w(throwable: Throwable) {
        forestAsArray.forEach {
            it.w(throwable)
        }
    }

    override fun e(message: String, vararg args: Any) {
        forestAsArray.forEach {
            it.e(message, args)
        }
    }

    override fun e(throwable: Throwable, message: String, vararg args: Any) {
        forestAsArray.forEach {
            it.e(throwable, message, args)
        }
    }

    override fun e(throwable: Throwable) {
        forestAsArray.forEach {
            it.e(throwable)
        }
    }

    override fun wtf(message: String, vararg args: Any) {
        forestAsArray.forEach {
            it.wtf(message, args)
        }
    }

    override fun wtf(throwable: Throwable, message: String, vararg args: Any) {
        forestAsArray.forEach {
            it.wtf(throwable, message, args)
        }
    }

    override fun wtf(throwable: Throwable) {
        forestAsArray.forEach {
            it.wtf(throwable)
        }
    }

    override fun log(priority: Int, message: String, vararg args: Any) {
        forestAsArray.forEach {
            it.log(priority, message, args)
        }
    }

    override fun log(priority: Int, throwable: Throwable, message: String, vararg args: Any) {
        forestAsArray.forEach {
            it.log(priority, throwable, message, args)
        }
    }

    override fun log(priority: Int, throwable: Throwable) {
        forestAsArray.forEach {
            it.log(priority, throwable)
        }
    }

    override fun onLog(priority: Int, tag: String, message: String, throwable: Throwable?) {
        //nothing
    }
}